### PR TITLE
Bugfix in broadcast_indexes

### LINKED ISSUE #1688: NotImplementedError: Vectorized indexing for <class 'xarray.core.indexing.LazilyIndexedArray'> is not implemented.
I think this is a regression in the current 0.10.0rc1:

#### Code Sample

```python
import xarray as xr
ds = xr.open_dataset('cesm_data.nc', decode_cf=False)
ds.temp.isel(time=ds.time < 274383)  # throws an error
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-18-a5c4179cd02d> in <module>()
----> 1 ds.temp.isel(time=ds.time < 274383)

~/.pyvirtualenvs/py3/lib/python3.5/site-packages/xarray/core/dataarray.py in isel(self, drop, **indexers)
    717         DataArray.sel
    718         """
--> 719         ds = self._to_temp_dataset().isel(drop=drop, **indexers)
    720         return self._from_temp_dataset(ds)
    721 

~/.pyvirtualenvs/py3/lib/python3.5/site-packages/xarray/core/dataset.py in isel(self, drop, **indexers)
   1278         for name, var in iteritems(self._variables):
   1279             var_indexers = {k: v for k, v in indexers_list if k in var.dims}
-> 1280             new_var = var.isel(**var_indexers)
   1281             if not (drop and name in var_indexers):
   1282                 variables[name] = new_var

~/.pyvirtualenvs/py3/lib/python3.5/site-packages/xarray/core/variable.py in isel(self, **indexers)
    771             if dim in indexers:
    772                 key[i] = indexers[dim]
--> 773         return self[tuple(key)]
    774 
    775     def squeeze(self, dim=None):

~/.pyvirtualenvs/py3/lib/python3.5/site-packages/xarray/core/variable.py in __getitem__(self, key)
    595         """
    596         dims, index_tuple, new_order = self._broadcast_indexes(key)
--> 597         data = self._indexable_data[index_tuple]
    598         if new_order:
    599             data = np.moveaxis(data, range(len(new_order)), new_order)

~/.pyvirtualenvs/py3/lib/python3.5/site-packages/xarray/core/indexing.py in __getitem__(self, key)
    414 
    415     def __getitem__(self, key):
--> 416         return type(self)(_wrap_numpy_scalars(self.array[key]))
    417 
    418     def __setitem__(self, key, value):

~/.pyvirtualenvs/py3/lib/python3.5/site-packages/xarray/core/indexing.py in __getitem__(self, key)
    394 
    395     def __getitem__(self, key):
--> 396         return type(self)(_wrap_numpy_scalars(self.array[key]))
    397 
    398     def __setitem__(self, key, value):

~/.pyvirtualenvs/py3/lib/python3.5/site-packages/xarray/core/indexing.py in __getitem__(self, key)
    361 
    362     def __getitem__(self, key):
--> 363         return type(self)(self.array, self._updated_key(key))
    364 
    365     def __setitem__(self, key, value):

~/.pyvirtualenvs/py3/lib/python3.5/site-packages/xarray/core/indexing.py in _updated_key(self, new_key)
    336             raise NotImplementedError(
    337                 'Vectorized indexing for {} is not implemented. Load your '
--> 338                 'data first with .load() or .compute().'.format(type(self)))
    339         new_key = iter(expanded_indexer(new_key, self.ndim))
    340         key = []

NotImplementedError: Vectorized indexing for <class 'xarray.core.indexing.LazilyIndexedArray'> is not implemented. Load your data first with .load() or .compute().

```
Here is the file: 
[cesm_data.nc.zip](https://github.com/pydata/xarray/files/1441729/cesm_data.nc.zip)

#### Expected Output

This used to work in v0.9

#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.10.0-38-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

xarray: 0.10.0rc1
pandas: 0.21.0
numpy: 1.13.3
scipy: 1.0.0
netCDF4: 1.3.0
h5netcdf: None
Nio: None
bottleneck: 1.2.1
cyordereddict: None
dask: 0.15.4
matplotlib: 2.1.0
cartopy: 0.15.1
seaborn: 0.8.1
setuptools: 36.6.0
pip: 9.0.1
conda: None
pytest: 3.2.3
IPython: 6.2.1
sphinx: 1.6.5
</details>

### PR DESCRIPTION
**Pull request by fujiisoup**
- [x] Closes #1688, #1694
 - [x] Tests added / passed
 - [x] Passes ``git diff upstream/master **/*py | flake8 --diff``
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

Fixes #1688.
It is caused that `Variable._broadcast_indexes` returns a wrong type of `Indexer`.
Now it supports the orthogonal-indexing with `LazilyIndexedArray`.

### CODE REVIEW

**Comment by fujiisoup** [comment]
I thought issue #1688 was already covered in `test_backend.py` but was not.
If someone could suggest an appropriate test in `test_backend.py`, it would be super nice.

**Comment by fmaussion** [comment]
This code reproduces the error (one can probably clean it up a little):

```
import xarray as xr
import numpy as np
a = np.arange(1e5)
da.to_netcdf('test.nc')
t = xr.open_dataarray('test.nc')
t.isel(t=t.t > 12)
```

File: xarray/tests/test_backends.py [hunk 1]
```diff
@@ -399,6 +399,14 @@ def test_orthogonal_indexing(self):
             actual = on_disk.isel(**indexers)
             self.assertDatasetIdentical(expected, actual)
 
+    def test_orthogonal_isel(self):
```
Comment (fujiisoup) [inline xarray/tests/test_backends.py]: @fmaussion 

Thanks. Added a high-level test here.

File: xarray/tests/test_backends.py [hunk 2]
```diff
@@ -1673,6 +1681,10 @@ def test_orthogonal_indexing(self):
         # pynio also does not support list-like indexing
         pass
 
+    def test_orthogonal_isel(self):
+        # pynio also does not support list-like indexing
```
Comment (shoyer) [inline xarray/tests/test_backends.py]: do we raise `NotImplementedError` with pynio when anything other than `BasicIndexer` is used?

File: xarray/tests/test_variable.py [hunk 3]
```diff
@@ -938,6 +940,39 @@ def test_detect_indexer_type(self):
         _, ind, _ = v._broadcast_indexes((vind, 3))
         assert type(ind) == indexing.VectorizedIndexer
 
+    def test_indexer_type(self):
```
Comment (shoyer) [inline xarray/tests/test_variable.py]: This is great, thanks!

File: xarray/backends/pynio_.py [hunk 4]
```diff
@@ -28,6 +29,17 @@ def get_array(self):
         return self.datastore.ds.variables[self.variable_name]
 
     def __getitem__(self, key):
+        if isinstance(key, indexing.VectorizedIndexer):
```
Comment (fujiisoup) [inline xarray/backends/pynio_.py]: @shoyer 
Thanks for your suggestion. Now it raises a NotImplementedError.
This backend seems to partly support outer indexing.

File: xarray/backends/pynio_.py [hunk 5]
```diff
@@ -28,6 +29,17 @@ def get_array(self):
...
+            raise NotImplementedError(
+                'Nio backend does not support vectorized indexing. Load your '
+                'data first with .load() or .compute(). Given {}'.format(key))
+        # Nio only supports slice indexing, not an array
+        if (isinstance(key, indexing.OuterIndexer) and
+                not all(isinstance(k, (integer_types, slice)) for k in key)):
```
Comment (shoyer) [inline xarray/backends/pynio_.py]: We should probably clarify on the indexer classes, but my understanding was:

- `BasicIndexer`: every element is an integer or slice
- `OuterIndexer`: every element is an integer, slice or 1d integer array
- `VectorizedIndexer`: every element is an integer, slice or N-dimensional integer array

Both outer and vectorized indexing generalize basic indexing, but in different ways.

So I think for this case we should already only be using a basic indexer.

Comment (fujiisoup) [reply xarray/backends/pynio_.py]: You are right.
I don't remember why I did that. Updated.

**Comment by fujiisoup** [comment]
Found another bug. Not ready yet.

**Comment by fujiisoup** [comment]
I found and fixed the following bugs. 

1. Fix for #1579 was not perfect (indexing `dask` array by `np.int` fails).
2. `LazilyIndexingArray` always used `OuterIndexer`.

**Comment by shoyer** [comment]
@fujiisoup this needs a merge from master after your last PR.

**Comment by fujiisoup** [comment]
@shoyer 
Thanks. Merged from the master.

But I think I should also add another test for #1694 , which seems to be solved by this PR.

**Comment by fujiisoup** [comment]
Found some an unfixed bug in #1676, where in some cases `MemoryCachedArray`, `CopyOnWriteArray`, `LazilyIndexedArray` has a bare `np.ndarray`, rather than `NumpyIndexingArray`.

I think this is now fixed.

File: xarray/tests/test_backends.py [hunk 6]
```diff
@@ -422,6 +423,47 @@ def test_vectorized_indexing(self):
...
+                        assert isinstance(obj, indexing.DaskIndexingAdapter)
+                    elif isinstance(obj.array, pd.Index):
+                        assert isinstance(obj, indexing.PandasIndexAdapter)
+        for k, v in ds.variables.items():
+            find_and_validate_array(v._data)
+
+    def test_array_type_after_indexing(self):
+        in_memory = create_test_data()
+        with self.roundtrip(in_memory) as on_disk:
+            self.validate_array_type(on_disk)
```
Comment (fujiisoup) [inline xarray/tests/test_backends.py]: This line makes sure that only `NumpyIndexingAdapter` stores `np.ndarray`, and other array wrappers don't.

File: xarray/tests/test_backends.py [hunk 7]
```diff
@@ -430,6 +431,39 @@ def test_isel_dataarray(self):
...
+            if hasattr(obj, 'array'):
+                if isinstance(obj.array, indexing.NDArrayIndexable):
+                    find_and_validate_array(obj.array)
+                else:
+                    if isinstance(obj.array, np.ndarray):
+                        assert isinstance(obj, indexing.NumpyIndexingAdapter)
+                    elif isinstance(obj.array, dask_array_type):
+                        assert isinstance(obj, indexing.DaskIndexingAdapter)
+                    elif isinstance(obj.array, pd.Index):
+                        assert isinstance(obj, indexing.PandasIndexAdapter)
```
Comment (shoyer) [inline xarray/tests/test_backends.py]: Should there be an error if none of these condition match?

**Comment by shoyer** [comment]
Thanks @fujiisoup, this is very helpful.

**Pull request merged**